gem/optimise.py

```diff
@@ -413,3 +480,555 @@ def aggressive_unroll(expression):
+    3. key = j, factors indexed with j, possibly also indexed with other indices (e.g. i)
+    4. key = k, factors indexed with k, possibly also indexed with other indices (e.g. i)
+    5. key = 2, factors indexed with more than 1 argument indices
+    This generalises to the cases with more than 2 argument indices
+    :param node: root of expression
+    :param argument_indices: tuple of argument (linear) indices
+    :return: tuple of dictionary of list of factors
+    """
+    monos = collect_terms(node, Sum)
+    result = []
+    arg_ind_set = set(argument_indices)
+    for mono in monos:
+        d = OrderedDict()
+        for i in [0, 1, 2] + list(argument_indices):
+            d[i] = list()
+        for factor in collect_terms(mono, Product):
+            fi = factor.free_indices
+            if not fi:
+                # no free indices
+                d[0].append(factor)
+            else:
+                ind_set = set(fi) & arg_ind_set
+                if len(ind_set) > 1:
+                    # more than 1 argument indices
+                    d[2].append(factor)
+                elif len(ind_set) == 0:
+                    # no argument indices
+                    d[1].append(factor)
+                else:
+                    # single argument index
+                    d[ind_set.pop()].append(factor)
+        for i in d:
+            d[i] = tuple(d[i])
+        result.append(d)
+    return tuple(result)
+
+
+def sumproduct_2_node(factor_lists):
+    """
+    generate gem node from list of factors
+    use recursion so that each term is not too long
+    i.e. (a+b) + (c+d) instead of (((a+b)+c)+d
+    :param factor_lists: list of factors representing sums of products
+    :return: gem node
+    """
+    if len(factor_lists) < 3:
+        sums = []
+        for fl in factor_lists:
+            sums.append(reduce(Product, fl, one))
+        return reduce(Sum, sums, Zero())
+    else:
+        mid = int(len(factor_lists) / 2)
+        return Sum(sumproduct_2_node(factor_lists[:mid]), sumproduct_2_node(factor_lists[mid:]))
+
+
+def factorise(factor_lists, quad_ind, arg_ind_flat):
+    """
+    recursively pick the most common factor
+    maybe can Memoize this one
+    :param factors: list of list of factors, representing sum of products
+    :return: optimised list of list of factors
+    """
+    if len(factor_lists) <= 1:
+        return factor_lists
+    # count number of common factors
+    counter = OrderedDict()
+    for fl in factor_lists:
+        fl_set = OrderedDict.fromkeys(fl)
+        if len(fl_set) > 1:
+            # at least two terms in product
+            for factor in fl_set:
+                if factor in counter:
+                    counter[factor] += 1
+                else:
+                    counter[factor] = 1
+
+    if not counter:
+        return factor_lists
+    # find which factor to factorise out first
+    mcf_value = [0, 1]  # (num arg indices, count)
```

Comment:
If you find yourself commenting your data structures like this, it's a smell that you've done something "bad".